htlcswitch/link.go

```diff
@@ -1294,18 +1322,35 @@ func (l *channelLink) htlcManager() {
 			}
 
 		case req := <-l.shutdownRequest:
-			// If the channel is clean, we send nil on the err chan
-			// and return to prevent the htlcManager goroutine from
-			// processing any more updates. The full link shutdown
-			// will be triggered by RemoveLink in the peer.
-			if l.channel.IsChannelClean() {
-				req.err <- nil
-				return
+			// This is a local shutdown request, so we set the
+			// shutdownInit bool and wait until there are no more
+			// updates to send. We should not forward HTLC's to our
+			// peer. Only update_fail, update_fulfill, and
+			// update_fee messages are allowed at this point.
+			l.shutdownInit = true
+			l.localCloseReq = req.closeReq
+
+			// Determine if upfront shutdown has been violated. If
+			// it hasn't, replace the DeliveryAddr with the
+			// user-provided one.
+			deliveryAddr, err := chooseDeliveryScript(
+				l.channel.LocalUpfrontShutdownScript(),
+				req.closeReq.DeliveryScript,
+			)
+			if err != nil {
+				// Send the error back so that the caller can
+				// be notified of the mismatch. The errChan is
+				// buffered.
+				req.errChan <- err
+				continue
+			}
+
+			if len(deliveryAddr) != 0 {
+				l.cfg.DeliveryAddr = deliveryAddr
```

Comment:
Don't think this should be here, but if it's just stashing, then it can live in the `channelLink` struct instead of the main config (this isn't externally set).